PadicsLoco/PadicsLoco.lean

```diff
@@ -0,0 +1,89 @@
+/-
+Copyright (c) 2024 Jou Glasheen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jou Glasheen, Kevin Buzzard
+-/
+
+import Mathlib.NumberTheory.Padics.RingHoms
+import Mathlib.Topology.MetricSpace.Ultra.Basic
+
+/-!
+# Local compactness of the p-adic numbers
+
+In this file, we prove that `Z_[p]` is totally bounded and compact,
+and that `ℚ_[p]` is locally compact.
+
+## Main results
+
+- `PadicInt.totally_bounded` : The set of p-adic integers `ℤ_[p]` is totally bounded.
+- `PadicInt.is_compact` : The set of p-adic integers `ℤ_[p]` is a compact topological space.
+- `Padic.locally_compact_space` : The set of p-adic numbers ℚ_[p] is a locally compact topological
+                                  space.
+
+## Notation
+
+ - `p` : Is a natural prime.
+
+## References
+
+Gouvêa, F. Q. (2020) p-adic Numbers An Introduction. 3rd edition.
+  Cham, Springer International Publishing
+-/
+open Metric Topology
+
+variable (p : ℕ) [Fact (Nat.Prime p)]
+
+namespace PadicInt
+
+/-- The set of p-adic integers `ℤ_[p]` is totally bounded. -/
+theorem totallyBounded_univ : TotallyBounded (univ : Set ℤ_[p]) := by
+  refine Metric.totallyBounded_iff.mpr (fun ε hε ↦ ?_)
+  obtain ⟨k, hk⟩ := exists_pow_neg_lt p hε
+  refine ⟨Nat.cast '' Finset.range (p ^ k), Set.toFinite _, fun z _ ↦ ?_⟩
+  simp only [PadicInt, Set.mem_iUnion, Metric.mem_ball, exists_prop, Set.exists_mem_image]
+  refine ⟨z.appr k, ?_, ?_⟩
+  · simpa only [Finset.mem_coe, Finset.mem_range] using z.appr_lt k
+  · exact (((z - z.appr k).norm_le_pow_iff_mem_span_pow k).mpr (z.appr_spec k)).trans_lt hk
+
+/-- The set of p-adic integers `ℤ_[p]` is a compact topological space. -/
+instance compactSpace : CompactSpace ℤ_[p] := by
+  rw [← isCompact_univ_iff, isCompact_iff_totallyBounded_isComplete]
+  exact ⟨totallyBounded_univ p, complete_univ⟩
+
+open Metric
+
+lemma closed_ball_at_zero : (closedBall 0 1) = {x : ℚ_[p] | ‖x‖ ≤ 1} := by
```

Comment:
This is true in any normed group, and we already have [mem_closedBall_zero_iff](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Normed/Group/Basic.html#mem_closedBall_zero_iff)